keps/draft-20190516-KUDO-package-manager.md

```diff
@@ -0,0 +1,159 @@
+  - TBD
+  - "@fabianbaier"
+editor: TBD
+creation-date: 2019-05-16
+status: provisional
+---
+
+# package-manager-for-distributing-kudo-packages
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a KEP and for highlighting any additional information provided beyond the standard KEP template.
+[Tools for generating][] a table of contents from markdown are available.
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+ * [Goals](#goals)
+ * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+ * [User Stories](#user-stories)
+    * [Framework Developer](#framework-developer)
+    * [Cluster Administrator](#cluster-administrator)
+ * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+ * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Infrastructure Needed](#infrastructure-needed)
+
+## Summary
+
+As more and more developers seed our `kubebuilder/frameworks` repository with incredibly important Frameworks, there needs to be a provided structure that talks about how this repository is organized and made available. The underlying and agreed on structure also indirectly impacts or tangents to other KEPs, e.g. [KEP-0003 CLI](https://github.com/kudobuilder/kudo/blob/master/keps/0003-kep-cli.md), [KEP-0007 CLI Gernation](https://github.com/kudobuilder/kudo/blob/master/keps/0007-cli-generation.md) or [KEP-0008 Framework Testing](https://github.com/kudobuilder/kudo/blob/master/keps/0008-framework-testing.md).
+
+Overall, this KEP should capture how we plan to provide a great user experience of installing Frameworks with KUDO. Attention should be split to:
+
+* How the Package structure looks like and dictates the overall repository structure
+* What tools should be used to achieve this structure
+* How are packages are distributed to the user
+
+## Motivation
+
+There are multiple package managers and ideas out there and we need to decide which concept will work the best for KUDO in its current state. This is not an Install CLI subcategory KEP, but of course we need to think about the repercussions of decisions made in this KEP. There are multiple stakeholders directly being impacted by decisions made here, but this also paths the way to think about new problems we haven't articulated yet (e.g. how to provide KUDO Frameworks in air-gapped clusters, verification of Frameworks and their tarbundles, etc.).
+
+### Goals
+
+Specific goals are:
+* Coming up with a structure for a Package
+* Defining the overall `kubebuilder/framework` repository structure
+* Having a solution on how to distribute this structure
+* Having this platform-agnostic so Packages can be distributed across multiple solutions
+* Adapting concepts from other successfull open-source projects in this regard
+
+### Non-Goals
+
+Non-goals are:
+* Re-inventing the wheel
```

Comment:
I agree, I believe we should follow a similar approach to https://github.com/kubernetes-sigs/krew/ and krew-index. Therefore, we should be able to package frameworks or make references to `custom` or external framework implementations via Kudo. 

As an example, I believe it'd be really beneficial to be able to refer to a tarball with all the required scripts (YAML or CUE or whatever is necessary to install the fwk) inside of it.
```
apiVersion: kudo.k8s.io/v1alpha1
kind: KudoFramework
metadata:
  name: debug-operator
spec:
  framework:
  - uri: https://github.com/kubebuilder/framework/kafka/archive/v1.0.2.zip
    sha256: 93c01f4b4f0862409dfbb10cb48995b70540f359e08742f0f10f7adf9d4326bd
```

Reply:
I like the idea of having one single yaml from which you can point to other frameworks. This is essentially also with what I POCed here https://github.com/fabianbaier/kudo/blob/cli-improvement/pkg/kudoctl/util/repo/repo.go

Take for example 

```
$ curl https://kudo-test-repo.storage.googleapis.com/index.yaml
apiVersion: v1
entries:
  kafka:
  - apiVersion: v1
    appVersion: 2.4.0
    name: kafka
    urls:
    - https://kudo-test-repo.storage.googleapis.com/kafka-0.1.0.tgz
    version: 0.1.0
  zookeeper:
  - apiVersion: v1
    appVersion: 3.4.10
    name: zookeeper
    urls:
    - https://kudo-test-repo.storage.googleapis.com/zookeeper-0.1.0.tgz
    version: 0.1.0
generated: 2019-04-29T21:56:54.555876047Z
```

In this `index.yaml` `urls` could link to internal or external resources to be downloaded. Where a good pattern would be by default having it within the same prefix of where the `index.yaml` lies (in this case  `https://kudo-test-repo.storage.googleapis.com/`) so you could also just by default say `kafka-0.1.0.tgz` and it would try to access it within the scope of the prefix, but then when giving a full valid url it also would "link" to .tgz files living outside the scope.

I am not sure if we need the `sha256` yet or if this can wait. For now I always assumed having "https" should be fine for now and just if demands for this would come up we can start thinking of this (and then looking at `go-tuf` or `helm` is doing it). WDYT?

Reply:
Just thinking about the external URLs, do we really need it? In the stable repo we should have stuff we kind of control and has a certain quality. If you allow linking to any external URL, how do you verify the content did not change? We would have to definitely come up with verification mechanism then.

Also what is the use case for that? Why would I add something external to the stable repo? Why wouldn't I want to host it in the repo as other?

I find this similar case as e.g. Android has an official AppStore with verified apps. Those are scanned and provide some level of security. If you feel like you want to go wild and want to install "something from the internet" you could do that by directly going to that website. In KUDO world that would be `kudo install http://my-hopefully-secured-site.com/framework.tgz` You don't need to have that in the official repo.

Reply:
ok I take this back, I took a closer look at krew... I think if we introduce the verification (hashing in their case) we could easily follow similar pattern and we don't have to maintain any storage (GCS). It has downsides like people can put whatever HTTP address there and when it's unavailable, they will blame potentially kudo? or when it's slow... with GCS we kind of guarantee some availability

Reply:
@alenkacz what if we verified this on every change (or periodically as well) and added an Event/status conditions to give feedback on these errors?

Reply:
> I like the idea of having one single yaml from which you can point to other frameworks. This is essentially also with what I POCed here https://github.com/fabianbaier/kudo/blob/cli-improvement/pkg/kudoctl/util/repo/repo.go

Good to see that ;)